packages/core/nano/src/types.ts

```diff
@@ -1,22 +1,139 @@
-export interface Consumer<T> {
-    notify: <K extends keyof T>(property: K, value: T[K]) => void;
-    notifyMany: <K extends keyof T>(values: Pick<T, K>) => void;
+export interface Pending<T> {
+  kind: "Pending";
+  estimate?: T;
 }
 
-export interface Pending<T> {
-    kind: "Pending";
-    estimate?: T;
+/**
+ * The interface for an object whose data can be bound to. We use this contract for
+ * components that want to expose their data and its changes to other components.
+ * An example of this would be Formula binding to a Tablero. In this scenario,
+ * Tablero is expected to implement IProducer and expose some of its data, given
+ * a reference, through getDataResult. Formula is expected to implement IConsumer
+ * and register itself as an observer to Tablero.
+ *
+ * Any component that implements IProducer is expected to provide some registration
+ * functionality and to notify observers whenever the data they are bound to changes.
+ */
+export interface IProducer<T> {
+  /**
+   * Binds the given observer to the component associated with this Producer.
+   * @param consumer - The object to start listening to any changes in the bound data.
+   */
+  addConsumer(consumer: IConsumer<T>): void;
+
+  /**
+   * Unbinds the given observer from the component associated with this Producer.
+   * @param consumer - The observer to unregister from the Producer.
+   */
+  removeConsumer(consumer: IConsumer<T>): void;
+
+  /**
+   * Return IDataResult for the given dataBindingReference, containing the data value
+   * and some additional metadata.
+   * @param dataBindingReference - An instance of dataBindingReference, a reference
+   * to some data in the bound component.
+   */
+  read<K extends keyof T>(consumer: IConsumer<T>, property: K, ...args: any[]): T[K] | Pending<T[K]>;
```

Comment:
I guess the semantics are that `read` subscribes only to the given property (merging with any existing subscriptions on the producer), and that `addConsumer` subscribes to everything?

Reply:
No, I just goofed.  I copied your signature for 'request()' without reconciling it with Bohemia's explicit subscription.

In my experience, it's almost always less expensive for the consumer to filter notifications than it is for the producer to track subscriptions to fine-grained properties, so I would recommend that contact always allows the producer the option of over-invalidating.

Of course, a producer can choose to be more frugal with its notifications, if it happens to be convenient.  We could also offer a generic IConsumer/IProducer that passes through the data from a source IProducer, but adds fine-grained dependency tracking to squelch notifications.

Reply:
Sounds good to me. So the contract of `read` is: get property `K` and subscribe to the producer. Subsequent notifications to the consumer will be bounded below by `K`, but possibly more.

Reply:
BTW - A cute trick for reducing the overhead of the implicit subscription model is to require the consumer to acquire a reader object from the producer, like this:
```ts
const reader = producer.open(consumer);
const length = reader.read('length');
for (let i = 0; i < length; i++) {
    a[i] = reader.read(i);
}
```
The producer adds the consumer to its set one time in 'open()' and then returns the same/cached reader object to every consumer.

You can deter consumers from caching readers by expiring them at the end of a recalc cycle, and mixing the allocation.

Reply:
Cool! In the implicit subscription model do we want per property invalidation in `notify`? Should it be the changed subset of all properties, or just the entire producer/reader and we let the consumer figure out what changed?

Reply:
And would the open/reader approach be part of the base interface?

Reply:
What's worked well for me in the past is to have the Producer call the consumer with each changed key

Reply:
So as it is now? Works for me